pkg/lib/controller-runtime/source/dynamic.go

```diff
@@ -0,0 +1,196 @@
+				log.Error(err, "failed to start dynamic source", "gvk", gvk)
+			}
+		}
+
+	}, 15*time.Second, d.stop)
+
+	synced.Wait()
+
+	return nil
+}
+
+var (
+	knownInformables []schema.GroupVersionKind
+	informablesLock  sync.RWMutex
+)
+
+// InformableGVKs returns the set of GVKs available in discovery that support the list and watch verbs (informable).
+func (d *Dynamic) InformableGVKs() (informables []schema.GroupVersionKind, err error) {
+	if d.client == nil {
+		return nil, fmt.Errorf("no client")
+	}
+
+	var resourceLists []*metav1.APIResourceList
+	resourceLists, err = d.client.ServerPreferredResources()
+	if err != nil {
+		informablesLock.RLock()
+		defer informablesLock.RUnlock()
+		log.Error(err, "could not get preferred resources, returning latest known gvks")
+		return knownInformables, nil
+	}
+
+	latest := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"list", "watch"}}, resourceLists)
+	for _, rl := range latest {
+		gv, err := schema.ParseGroupVersion(rl.GroupVersion)
+		if err != nil {
+			log.Error(err, "could not parse group version")
+			continue
+		}
+
+		for _, r := range rl.APIResources {
+			gvk := schema.GroupVersionKind{Group: gv.Group, Version: gv.Version, Kind: r.Kind}
+			informables = append(informables, gvk)
+		}
+	}
+
+	if len(informables) > 0 {
+		// Memoize the latest informables in case discovery fails on successive calls
+		informablesLock.Lock()
+		defer informablesLock.Unlock()
+		knownInformables = informables
+	}
+
+	return
+}
+
+func (d *Dynamic) startSource(gvk schema.GroupVersionKind, handle handler.EventHandler, queue workqueue.RateLimitingInterface, predicates ...predicate.Predicate) (err error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.sources == nil {
+		d.sources = map[schema.GroupVersionKind]source.Source{}
+	}
+
+	r := &unstructured.Unstructured{}
+	r.SetGroupVersionKind(gvk)
+	s := &source.Kind{Type: r}
+	if err = s.InjectCache(d.cache); err != nil {
+		return
+	}
+
+	if err = s.Start(handle, queue, predicates...); err != nil {
+		return
+	}
+
+	log.Info("dynamic source started", "gvk", gvk)
+
+	d.sources[gvk] = s
+
+	return
+}
+
+func (d *Dynamic) sourceStarted(gvk schema.GroupVersionKind) (added bool) {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	_, added = d.sources[gvk]
+
+	return
+}
+
+// InjectCache injects the Cache dependency.
+// It is primarily used by the ControllerManager to share a Cache between Sources.
+func (d *Dynamic) InjectCache(c ctrlcache.Cache) error {
+	if d.cache == nil {
+		d.cache = c
+	}
+	return nil
+}
+
+// InjectStopChannel is internal should be called only by the Controller.
+// It is used to inject the stop channel initialized by the ControllerManager.
+func (d *Dynamic) InjectStopChannel(stop <-chan struct{}) error {
+	if d.stop == nil {
+		d.stop = stop
+	}
+
+	return nil
+}
+
+// InjectConfig injects the the REST Config dependency.
+func (d *Dynamic) InjectConfig(config *rest.Config) (err error) {
```

Comment:
Is `InjectRESTConfigDependency` too big of a name? It'll would get rid of the comment, and be self explanatory in the calling site too.

Reply:
I don't disagree, but we don't control the [inject method signatures](https://github.com/kubernetes-sigs/controller-runtime/blob/49e1044b3b6ac6b3064cd348604e5b28e9d5c786/pkg/runtime/inject/inject.go#L61).

Reply:
> It'll would get rid of the comment

Also, golangci-lint generally complains about exports w/o comments.

Reply:
Alas.